### ISSUE TITLE
Metadata import

### ISSUE THREAD

**Issue by JaimieMurdock** [inpho/topic-explorer#104]
Create a `vsm metadata` command that:
1. Generates a csv file with the first column of ids
2. Imports a csv file with first column of labels, second column of new datums
3. Add support for headers to specify metadata object names
4. Add support for redoing labels
5. Add support to `vsm serve` for using alternate metadata fields for title display.

**Comment by JaimieMurdock** [comment]
This is in progress at branch `metadata` and specifically in [metadata.py](https://github.com/inpho/topic-explorer/blob/bb6416a9c0c06e17c4d297df6163f3eddb6e8fed/topicexplorer/metadata.py)

**Comment by JaimieMurdock** [comment]
I've merged in the first step from the `metadata` branch. 

IDs can be listed via `topicexplorer metadata -l filename`.
New metadata can be imported via `topicexplorer metadata -a filename`.

Headers in the new metadata file specify the corpus fields.

I have not yet added:
1. Support for renaming items
2. Support for alternate metadata fields for title display.

The alternate display is somewhat sketched in `topicexplorer.extensions.title` - what is needed is documentation for a new config section that would allow an arbitrary field to be selected.

**Comment by JaimieMurdock** [comment]
This has been merged into the `master` branch as of release 1.0b121. The support for renaming items is not yet complete. The aribtrary item labels have been split to #162

**Comment by JaimieMurdock** [comment]
One remaining problem with the metadata import: field sizes are not adjusted, so the fields do not change size - this is a huge problem related to how we store metadata as numpy arrays rather than as native dictionaries with arbitrary field size.

**Comment by JaimieMurdock** [comment]
One potential solution to the numpy string problem is to use `dtype='object'` rather than `dtype='S8'` to represent the strings. However, this introduces overhead of numpy working with Python object references.

The real question is whether a list of strings, numpy.array of strings or numpy.array of objects is more efficient... could be a nice blog post.

[python numpy array of arbitrary length strings](http://stackoverflow.com/questions/14639496/python-numpy-array-of-arbitrary-length-strings)

**Comment by JaimieMurdock** [comment]
For posterity's sake, here's some benchmarks:

```
import numpy
import random
import sys
c = c = list('a' * random.randint(0,20) for _ in range(10000))
a = numpy.array(c)
b = numpy.array(c, dtype=object)

%timeit a.copy()
print sys.getsizeof(a)

%timeit b.copy()
print sys.getsizeof(b)

%timeit c[:]
print sys.getsizeof(c)
```

```
100000 loops, best of 3: 14.9 µs per loop
200096
10000 loops, best of 3: 43.7 µs per loop
80096
10000 loops, best of 3: 46.9 µs per loop
83112
```

Note that the `S20` representation is 2.5 times as large as the native list and object array are roughly equivalent in size. The object array is smaller than the native list because it pre-allocates the space and doesn't have to maintain the linked list representation.

**Comment by JaimieMurdock** [comment]
Quick way to convert all metadata fields

```python
ctx_dtype = v.corpus.context_data[0].dtype
[(t, object if ctx_dtype.fields[t][0].char in ['U', 'S', 'a'] else ctx_dtype.fields[t][0]) for t in ctx_dtype.names]
```

Testing size:
```python
import sys

print sys.getsizeof(v.corpus.words)
# 715720
a = v.corpus.words.astype(object)
b = sum([sys.getsizeof(w) for w in a])
print sys.getsizeof(a) + b
# 495444
```

**Comment by JaimieMurdock** [comment]
Fixed in vsm 1.0b6.
